### PR TITLE
Fix update_map_type/2 crash with a variable as the argument

### PR DESCRIPTION
**Pull request by erszcz**
This PR addresses #233 following @zuiderkwast's suggestions from https://github.com/josefs/Gradualizer/pull/233#issuecomment-548869362. EDIT: ~It's still a work in progress.~

Adding the `update_map_type({var, _, TyVar}, AssocTys) -> ...` clause fixes the crash. EDIT: ~I'm not sure what constraints or at which level should be added, though.~ I've removed returning constraints from `update_map_type`, but I'm leaving the rest of this message for clarity and reference.

---

The code in question is:

```erlang
-spec t(list()) -> map().
t(Rows) ->
    lists:foldl(fun(Id, List) ->
                        List#{id := Id}
                end, #{id => undefined}, Rows).
```

Let's have a look at the trace of `type_check_call` (I'm leaving the inner calls untouched, but the most important are the call to `type_check_call` and then the final return from it at the very bottom):

```erlang
     1	{{2019,12,17},{15,41,37}}
     2	{trace,<0.87.0>,call,
     3	    {typechecker,type_check_call,
     4	        [{env,
     5	             #{{t,1} =>
     6	                   [{type,
     7	                        {5,8},
     8	                        bounded_fun,
     9	                        [{type,
    10	                             {5,8},
    11	                             'fun',
    12	                             [{type,{5,8},product,[{type,{5,9},list,[]}]},
    13	                              {type,{5,20},map,any}]},
    14	                         []]}]},
    15	             #{},
    16	             #{'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
    17	             {tenv,update_map_type_error,#{},#{}},
    18	             false,false,true},
    19	         {type,{5,20},map,any},
    20	         {call,
    21	             {7,10},
    22	             {remote,{7,10},{atom,{7,5},lists},{atom,{7,11},foldl}},
    23	             [{'fun',
    24	                  {7,17},
    25	                  {clauses,
    26	                      [{clause,
    27	                           {7,20},
    28	                           [{var,{7,21},'Id'},{var,{7,25},'List'}],
    29	                           [],
    30	                           [{map,
    31	                                {8,29},
    32	                                {var,{8,25},'List'},
    33	                                [{map_field_exact,
    34	                                     {8,31},
    35	                                     {atom,{8,31},id},
    36	                                     {var,{8,37},'Id'}}]}]}]}},
    37	              {map,
    38	                  {9,22},
    39	                  [{map_field_assoc,
    40	                       {9,24},
    41	                       {atom,{9,24},id},
    42	                       {atom,{9,30},undefined}}]},
    43	              {var,{9,42},'Rows'}]},
    44	         {fun_ty,
    45	             [{type,25,'fun',
    46	                  [{type,25,product,[{var,25,'T'},{var,25,'Acc'}]},
    47	                   {var,25,'Acc'}]},
    48	              {var,25,'Acc'},
    49	              {type,25,list,[{var,25,'T'}]}],
    50	             {var,25,'Acc'},
    51	             {constraints,#{},#{},
    52	                 {set,0,16,16,8,80,48,
    53	                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
    54	                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}}}},
    55	         [{'fun',
    56	              {7,17},
    57	              {clauses,
    58	                  [{clause,
    59	                       {7,20},
    60	                       [{var,{7,21},'Id'},{var,{7,25},'List'}],
    61	                       [],
    62	                       [{map,
    63	                            {8,29},
    64	                            {var,{8,25},'List'},
    65	                            [{map_field_exact,
    66	                                 {8,31},
    67	                                 {atom,{8,31},id},
    68	                                 {var,{8,37},'Id'}}]}]}]}},
    69	          {map,
    70	              {9,22},
    71	              [{map_field_assoc,
    72	                   {9,24},
    73	                   {atom,{9,24},id},
    74	                   {atom,{9,30},undefined}}]},
    75	          {var,{9,42},'Rows'}],
    76	         {{7,10},
    77	          {remote,{7,10},{atom,{7,5},lists},{atom,{7,11},foldl}},
    78	          [{type,25,bounded_fun,
    79	               [{type,25,'fun',
    80	                    [{type,25,product,
    81	                         [{type,25,'fun',
    82	                              [{type,25,product,[{var,25,'T'},{var,25,'Acc'}]},
    83	                               {var,25,'Acc'}]},
    84	                          {var,25,'Acc'},
    85	                          {type,25,list,[{var,25,'T'}]}]},
    86	                     {var,25,'Acc'}]},
    87	                []]}]}]}}
      	
    88	{{2019,12,17},{15,41,37}}
    89	{trace,<0.87.0>,call,
    90	    {typechecker,type_check_expr_in,
    91	        [{env,
    92	             #{{t,1} =>
    93	                   [{type,
    94	                        {5,8},
    95	                        bounded_fun,
    96	                        [{type,
    97	                             {5,8},
    98	                             'fun',
    99	                             [{type,{5,8},product,[{type,{5,9},list,[]}]},
   100	                              {type,{5,20},map,any}]},
   101	                         []]}]},
   102	             #{},
   103	             #{'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   104	             {tenv,update_map_type_error,#{},#{}},
   105	             false,false,true},
   106	         {type,25,'fun',
   107	             [{type,25,product,[{var,25,'T'},{var,25,'Acc'}]},{var,25,'Acc'}]},
   108	         {'fun',
   109	             {7,17},
   110	             {clauses,
   111	                 [{clause,
   112	                      {7,20},
   113	                      [{var,{7,21},'Id'},{var,{7,25},'List'}],
   114	                      [],
   115	                      [{map,
   116	                           {8,29},
   117	                           {var,{8,25},'List'},
   118	                           [{map_field_exact,
   119	                                {8,31},
   120	                                {atom,{8,31},id},
   121	                                {var,{8,37},'Id'}}]}]}]}}]}}
      	
   122	{{2019,12,17},{15,41,37}}
   123	{trace,<0.87.0>,call,
   124	    {typechecker,type_check_expr_in,
   125	        [{env,
   126	             #{{t,1} =>
   127	                   [{type,
   128	                        {5,8},
   129	                        bounded_fun,
   130	                        [{type,
   131	                             {5,8},
   132	                             'fun',
   133	                             [{type,{5,8},product,[{type,{5,9},list,[]}]},
   134	                              {type,{5,20},map,any}]},
   135	                         []]}]},
   136	             #{},
   137	             #{'Id' => {var,25,'T'},
   138	               'List' => {var,25,'Acc'},
   139	               'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   140	             {tenv,update_map_type_error,#{},#{}},
   141	             false,false,true},
   142	         {var,25,'Acc'},
   143	         {map,
   144	             {8,29},
   145	             {var,{8,25},'List'},
   146	             [{map_field_exact,{8,31},{atom,{8,31},id},{var,{8,37},'Id'}}]}]}}
      	
   147	{{2019,12,17},{15,41,37}}
   148	{trace,<0.87.0>,call,
   149	       {typechecker,update_map_type,
   150	                    [{var,25,'Acc'},
   151	                     [{type,{8,31},
   152	                            map_field_exact,
   153	                            [{atom,{8,31},id},{var,25,'T'}]}]]}}
      	
   154	{{2019,12,17},{15,41,37}}
   155	{trace,<0.87.0>,return_from,
   156	       {typechecker,update_map_type,2},
   157	       {{type,0,any,[]},
   158	        {constraints,#{},#{},
   159	                     {set,0,16,16,8,80,48,
   160	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   161	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   162	                            []}}}}}}
      	
   163	{{2019,12,17},{15,41,37}}
   164	{trace,<0.87.0>,return_to,{typechecker,do_type_check_expr_in,3}}
      	
   165	{{2019,12,17},{15,41,37}}
   166	{trace,<0.87.0>,return_from,
   167	       {typechecker,type_check_expr_in,3},
   168	       {#{},
   169	        {constraints,#{},#{},
   170	                     {set,0,16,16,8,80,48,
   171	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   172	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   173	                            []}}}}}}
      	
   174	{{2019,12,17},{15,41,37}}
   175	{trace,<0.87.0>,return_to,{typechecker,check_clause,4}}
      	
   176	{{2019,12,17},{15,41,37}}
   177	{trace,<0.87.0>,return_from,
   178	       {typechecker,type_check_expr_in,3},
   179	       {#{'Id' => {var,25,'T'},
   180	          'List' => {var,25,'Acc'},
   181	          'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   182	        {constraints,#{},#{},
   183	                     {set,0,16,16,8,80,48,
   184	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   185	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   186	                            []}}}}}}
      	
   187	{{2019,12,17},{15,41,37}}
   188	{trace,<0.87.0>,return_to,{lists,zipwith,3}}
      	
   189	{{2019,12,17},{15,41,37}}
   190	{trace,<0.87.0>,call,
   191	    {typechecker,type_check_expr_in,
   192	        [{env,
   193	             #{{t,1} =>
   194	                   [{type,
   195	                        {5,8},
   196	                        bounded_fun,
   197	                        [{type,
   198	                             {5,8},
   199	                             'fun',
   200	                             [{type,{5,8},product,[{type,{5,9},list,[]}]},
   201	                              {type,{5,20},map,any}]},
   202	                         []]}]},
   203	             #{},
   204	             #{'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   205	             {tenv,update_map_type_error,#{},#{}},
   206	             false,false,true},
   207	         {var,25,'Acc'},
   208	         {map,
   209	             {9,22},
   210	             [{map_field_assoc,
   211	                  {9,24},
   212	                  {atom,{9,24},id},
   213	                  {atom,{9,30},undefined}}]}]}}
      	
   214	{{2019,12,17},{15,41,37}}
   215	{trace,<0.87.0>,call,
   216	       {typechecker,update_map_type,
   217	                    [{type,0,map,any},
   218	                     [{type,{9,24},
   219	                            map_field_assoc,
   220	                            [{atom,{9,24},id},{atom,{9,30},undefined}]}]]}}
      	
   221	{{2019,12,17},{15,41,37}}
   222	{trace,<0.87.0>,return_from,
   223	       {typechecker,update_map_type,2},
   224	       {{type,0,map,
   225	              [{type,{9,24},
   226	                     map_field_exact,
   227	                     [{atom,{9,24},id},{atom,{9,30},undefined}]},
   228	               {type,0,map_field_assoc,[{type,0,any,[]},{type,0,any,[]}]}]},
   229	        {constraints,#{},#{},
   230	                     {set,0,16,16,8,80,48,
   231	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   232	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   233	                            []}}}}}}
      	
   234	{{2019,12,17},{15,41,37}}
   235	{trace,<0.87.0>,return_to,{typechecker,do_type_check_expr_in,3}}
      	
   236	{{2019,12,17},{15,41,37}}
   237	{trace,<0.87.0>,return_from,
   238	       {typechecker,type_check_expr_in,3},
   239	       {#{},
   240	        {constraints,#{'Acc' => [{type,0,map,any}]},
   241	                     #{},
   242	                     {set,0,16,16,8,80,48,
   243	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   244	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   245	                            []}}}}}}
      	
   246	{{2019,12,17},{15,41,37}}
   247	{trace,<0.87.0>,return_to,{lists,zipwith,3}}
      	
   248	{{2019,12,17},{15,41,37}}
   249	{trace,<0.87.0>,call,
   250	    {typechecker,type_check_expr_in,
   251	        [{env,
   252	             #{{t,1} =>
   253	                   [{type,
   254	                        {5,8},
   255	                        bounded_fun,
   256	                        [{type,
   257	                             {5,8},
   258	                             'fun',
   259	                             [{type,{5,8},product,[{type,{5,9},list,[]}]},
   260	                              {type,{5,20},map,any}]},
   261	                         []]}]},
   262	             #{},
   263	             #{'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   264	             {tenv,update_map_type_error,#{},#{}},
   265	             false,false,true},
   266	         {type,25,list,[{var,25,'T'}]},
   267	         {var,{9,42},'Rows'}]}}
      	
   268	{{2019,12,17},{15,41,37}}
   269	{trace,<0.87.0>,return_from,
   270	       {typechecker,type_check_expr_in,3},
   271	       {#{},
   272	        {constraints,#{},#{},
   273	                     {set,0,16,16,8,80,48,
   274	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   275	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   276	                            []}}}}}}
      	
   277	{{2019,12,17},{15,41,37}}
   278	{trace,<0.87.0>,return_to,{lists,zipwith,3}}
      	
   279	{{2019,12,17},{15,41,37}}
   280	{trace,<0.87.0>,return_from,
   281	       {typechecker,type_check_call,6},
   282	       {#{'Id' => {var,25,'T'},
   283	          'List' => {var,25,'Acc'},
   284	          'Rows' => {type,{5,9},list,[{type,{5,9},any,[]}]}},
   285	        {constraints,#{'Acc' => [{type,0,map,any}]},
   286	                     #{'Acc' => [{type,0,map,any}]},
   287	                     {set,0,16,16,8,80,48,
   288	                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
   289	                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
   290	                            []}}}}}}
```

My hunch is that the lower bound constraint returned from `type_check_call` (line 285 above) should be a map with at least the `id => undefined` association. However, it might not be connected with the newly added `update_map_type` clause at all. If it's not the case, then I think we don't have to extend `update_map_type` with returning new constraints - I'd amend the relevant commit if so.

### CODE REVIEW

File: test/should_pass/update_map_type_error.erl [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-module(update_map_type_error).
+
+-compile(export_all).
+
+-spec t(list()) -> map().
+t(Rows) ->
+    lists:foldl(fun(Id, List) ->
+                        List#{id := Id}
```
Comment (zuiderkwast) [inline test/should_pass/update_map_type_error.erl]: `List` is a slightly confusing name for a map.

**Comment by erszcz** [comment]
I've removed returning constraints from `update_map_type`. Handling `{var, _, _}` is enough to fix the crash which was the original problem.

File: test/should_pass/update_map_type_error.erl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-module(update_map_type_error).
```
Comment (zuiderkwast) [inline test/should_pass/update_map_type_error.erl]: Now that the test passes, it shouldn't be named "error".

**Comment by zuiderkwast** [comment]
I believe the first argument to update_map_type/2 is actually a type (by looking at the other function clauses) and thus in the added clause, it's a type variable. In the provided trace, we can see that it's actually the type variable `Acc` from the spec of lists:foldl/3.

**Comment by erszcz** [comment]
@zuiderkwast Ahhh, indeed. You're right! I was obviously confused when committing the rename 😕

**Comment by zuiderkwast** [comment]
It doesn't matter. `{var, _, Var}` is fine for a type var as well.

**Pull request merged**